src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,261 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic that handles every state transition
+ */
+public class LabelPickerState {
+
+    private Set<String> initialLabels;
+    private List<String> addedLabels;
+    private List<String> removedLabels;
+    private List<String> matchedLabels;
+    private List<TurboLabel> allLabels;
+    private OptionalInt currentSuggestionIndex;
+
+    @SuppressWarnings("PMD")
```

Comment:
```java
@SuppressWarnings("PMD")
    private final void handleUserInput(ObservableValue<? extends String> observable, 
                                 String oldText, String newText) {
        // NO_PMD: method parameters are unused because the entire text will be processed 
        state = new LabelPickerState(
            new HashSet<>(issue.getLabels()), allLabels, queryField.getText().toLowerCase());
        populatePanes(state);
    }
```